fastai/core.py

```diff
@@ -3,37 +3,35 @@
 
 def sum_geom(a,r,n): return a*n if r==1 else math.ceil(a*(1-r**n)/(1-r))
 
+def listy(x): return isinstance(x, (list,tuple))
+def sapply_(x, f): return [f(o) for o in x] if listy(x) else f(x)
```

Comment:
That can be more generally useful, so maybe remove the trailing underscore. But try to come up with a name that's more understandable to python users, since sapply is an R thing. :)

Reply:
Also the conditional here should be `if isinstance(e, collections.Iterable)`, since you're not using the objects indexer so anything iterable will work.

Reply:
For naming ideas, this feels a lot like the FP notion of `map` e.g. as in Haskell. That's a reserved keyword in Python though, and the Java'y naming of `transform` is overloaded with image preprocessing. So perhaps `map_over` or `apply`? The latter is kind of generic, so I picked `map_over` for this volley.

Done on the `isinstance` change, extracted to a function used here. Seems like something for `collections.Mapping` might be nice as well.

Reply:
@jph00 If I get @mcskinner the whole point of this function is to apply a `f` on a pythonic structure of objects including `tensors` ,  `ndarrays`, `DataSets` (which are `Iterable`)

This seems to be very useful especially if we allow for nesting of structures, imagine:
```py
>>> ds = trn_ds, val_ds, tst_ds
>>> sapply(ds, DataLoader)    # -> (trn_dl, val_dl, tst_dl)

>>> sapply( (x, [mask, borders, weights]), crop)  #-> (corp(x), [crop(mask), crop(borders), crop(weights)])
```

Then the name `sapply` seems to  reflect the idea well as it can be read as "structure apply". Which explains why the function don't work on any`Iterable` and only on list tuples nad dicts and why `sapply(obj, f)`  is  `f(obj)` and not `map(obj, f)` (ie. because obj is not embedded in  a pythonic structure)

Such function can simplify  `get_tensor` from data loaders for example:
https://github.com/fastai/fastai/blob/master/fastai/dataloader.py#L10-L20

If we make sapply as follows:
```py
def sapply(x, f):
    if isinstance(x, collections.Mapping):
        return {k: sapply(sample, f) for k, sample in x.items()}
    elif isinstance(x, collections.List):
        return [sapply(sample, f) for sample in x]
    elif isinstance(x, collections.Tuple):
        return tuple(sapply(sample, f) for sample in x)
    else:
        return f(x)
```    
    
We can write:
```py
def pin_memory(batch):
    if isinstance(batch, (np.ndarray, np.generic)):
        batch = T(batch).contiguous()
        return batch.pin_memory()
    if isinstance(x, string_classes): return bach
    raise TypeError("Batch must contain be in structure of dicts, lists or tuples; found {}"
                     .format(type(batch)))

def get_tensor(batch, pin): return sapply(batch, pin_memory) if pin else batch
```

Reply:
I've just checked that `sapply` in `R` does not behave in the way described above, so it will be confusing for some. Maybe `struct_apply` then? It can be confused with https://docs.python.org/3/library/struct.html, but hopefully noone will mix DeepLearning with writing low level code to construct arbitrary `C` structures.